Save card token

DanielePalombo opened the pull request:
Save the card on Square in order to re-use it for the subscription.

Following the instruction specified on [Square Documentation](https://developer.squareup.com/docs/cards-api/walkthrough/card-from-payment-id) this PR does:

- Create the customer on Square when the payment is saved. ref https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/spree/payment.rb#L201
- If SolidusSquare::PaymentSource#token is empty, save the card to square and store its ID on the token ref https://github.com/nebulab/solidus_square/pull/69/files#diff-1bafe74d854b0b95c82a0db695d314b51be7e18315a0a990f400bf1df69b81fcR144-R147
- if SolidusSquare::PaymentSource#token is present, use token and customer_id to create the payment ref https://github.com/nebulab/solidus_square/pull/69/files#diff-1bafe74d854b0b95c82a0db695d314b51be7e18315a0a990f400bf1df69b81fcR137

I have QA it with:

both new and existing card
auto capture enabled and disabled
Guest user

<details>
  <summary>Details</summary>
  
  New card
<img width="722" alt="Screenshot 2022-01-20 at 16 38 34" src="https://user-images.githubusercontent.com/387690/150377020-0fdc904f-3b24-4b84-b500-c1c0819b0861.png">

Existing card
<img width="724" alt="Screenshot 2022-01-20 at 16 39 31" src="https://user-images.githubusercontent.com/387690/150377027-05d09bba-b21c-4a27-b8c7-c6ddc79c1e18.png">

Guest user
<img width="983" alt="Screenshot 2022-01-20 at 16 40 24" src="https://user-images.githubusercontent.com/387690/150377317-523118fe-f159-42a7-9ee6-1c00d687a506.png">

Auto capture false
<img width="1131" alt="Screenshot 2022-01-20 at 16 34 43" src="https://user-images.githubusercontent.com/387690/150377538-06246c01-8b93-45b6-8c30-90a32242d361.png">
<img width="1151" alt="Screenshot 2022-01-20 at 16 34 54" src="https://user-images.githubusercontent.com/387690/150377551-dd51b113-dc47-44c5-99cc-84c231645ee0.png">
</details>

DanielePalombo committed:
Add SolidusSquare::Customer

DanielePalombo committed:
Add customer to SolidusSquare::PaymentSource

DanielePalombo committed:
Add SolidusSquare::Gateway#create_profile

This method is called by solidus when the payment is created*.

The method creates the square_customer if it's not already present, and
stores the customer ID on the square_customer_ref.

created*
https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/spree/payment.rb#L28
https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/spree/payment.rb#L193

DanielePalombo committed:
Add related methods to PaymentMethod and Source

SolidusSquare::PaymentMethod#create_profile
SolidusSquare::PaymentMethod#payment_profiles_supported?
SolidusSquare::PaymentSource#reusable?

DanielePalombo committed:
Add SolidusSquare::Cards::Create

Add a service to create the card on square.

https://developer.squareup.com/reference/square/cards-api/create-card

DanielePalombo committed:
Add customer_id to the Payments::Create service

DanielePalombo committed:
Store the Square card

When the payment is created succesfully if the payment source doesn't
contain the token, save the card and store its id on the token, so it
can be used for further payment like subscription.

DanielePalombo committed:
Add the template to show existing square payments

DanielePalombo committed:
Fix the payment sync service

Right now when the webhook of square is received the customer profile
is created on square.
This commit updates the spec and use the env set on SolidusSquare

DanielePalombo committed:
Skip square card creation for guest user

MassimilianoLattanzio reviewed (approved):
LGTM

QA:
1. Create an order with a new user also create the customer on Square and save all data.
2. With returning users, we can reuse the old payment method.
3. I tried both with and without auto-capture, and it is working fine.
4. I also placed an order with a guest user. The payment/order was created correctly, and no customer was created on Square as expected.

Everything works as expected. Excellent work ❤️

DanielePalombo committed:
Refactor create_payment_on_square

Minor refactor of the method create_payment_on_square to make it more
clear and readable.

DanielePalombo committed:
Don't create the card if profile is not enabled

Don't create the card if
Spree::PaymentMethod::payment_profiles_supported? false

DanielePalombo commented:
@MassimilianoLattanzio the checks are all green now 🥳

kennyadsl reviewed (approved):
👍 Thanks Dani!
